release/releaser.go

```diff
@@ -44,233 +42,376 @@ func NewReleaser(
-	case params.ServiceSpec == flux.ServiceSpecAll && params.ImageSpec == flux.ImageSpecNone:
-		releaseType = "release_all_without_update"
-		actions, err = r.releaseWithoutUpdate(releaseType, msg, inst, services)
+	logStatus("Looking up images.")
+	if params.ImageSpec != flux.ImageSpecNone {
+		// Figure out how the services are to be updated.
+		if updates, err = r.calculateImageUpdates(rc, updates, &params, logStatus); err != nil {
+			return nil, err
+		}
+	}
 
-	case params.ServiceSpec == flux.ServiceSpecAll:
-		releaseType = "release_all_for_image"
-		actions, err = r.releaseImages(releaseType, msg, inst, services, images)
+	if params.Kind == flux.ReleaseKindExecute {
+		err = r.execute(rc, updates, &params, logStatus)
+	}
 
-	case params.ImageSpec == flux.ImageSpecLatest:
-		releaseType = "release_one_to_latest"
-		actions, err = r.releaseImages(releaseType, msg, inst, services, images)
+	return nil, err
+}
 
-	case params.ImageSpec == flux.ImageSpecNone:
-		releaseType = "release_one_without_update"
-		actions, err = r.releaseWithoutUpdate(releaseType, msg, inst, services)
+// Take the spec given in the job, and figure out which services are
+// in question based on the running services and those defined in the
+// repo. Fill in the release result along the way.
+func (r *Releaser) selectServices(rc *ReleaseContext, releaseJob *jobs.ReleaseJobParams, logStatus statusFn) ([]*ServiceUpdate, error) {
+	lockedSet, err := rc.LockedServices()
+	if err != nil {
+		return nil, err
+	}
 
+	excludedSet := flux.ServiceIDSet{}
+	excludedSet.Add(releaseJob.Excludes)
+
+	// For backwards-compatibility, there's two fields: ServiceSpec
+	// and ServiceSpecs. An entry in ServiceSpec takes precedence.
+	switch releaseJob.ServiceSpec {
+	case flux.ServiceSpec(""):
+		ids := []flux.ServiceID{}
+		for _, s := range releaseJob.ServiceSpecs {
+			if s == flux.ServiceSpecAll {
+				return rc.SelectServices(nil, lockedSet, excludedSet, logStatus)
+			}
+			id, err := flux.ParseServiceID(string(s))
+			if err != nil {
+				return nil, err
+			}
+			ids = append(ids, id)
+		}
+		return rc.SelectExactServices(ids, lockedSet, excludedSet, logStatus)
+	case flux.ServiceSpecAll:
+		return rc.SelectServices(nil, lockedSet, excludedSet, logStatus)
 	default:
-		releaseType = "release_one"
-		actions, err = r.releaseImages(releaseType, msg, inst, services, images)
+		id, err := flux.ParseServiceID(string(releaseJob.ServiceSpec))
+		if err != nil {
+			return nil, err
+		}
+		return rc.SelectExactServices([]flux.ServiceID{id}, lockedSet, excludedSet, logStatus)
 	}
-	return releaseType, actions, err
 }
 
-func (r *Releaser) releaseImages(method, msg string, inst *instance.Instance, getServices ServiceSelector, getImages ImageSelector) ([]ReleaseAction, error) {
-	var res []ReleaseAction
-	res = append(res, r.releaseActionPrintf(msg))
-
-	var (
-		base  = r.metrics.StageDuration.With("method", method)
-		stage *metrics.Timer
-	)
+// --- Additions to ReleaseContext for above
```

Comment:
Yep, just collecting them here until I've got all the ones I need, so less chance of leaving dead code around.